src/Interpreters/Cache/FileCache.cpp

```diff
@@ -1077,274 +897,288 @@ void FileCache::loadCacheInfoIntoMemory(std::lock_guard<std::mutex> & cache_lock
+    /// Shuffle file_segment_metadatas to have random order in LRUQueue as at startup all file_segment_metadatas have the same priority.
     pcg64 generator(randomSeed());
     std::shuffle(queue_entries.begin(), queue_entries.end(), generator);
-    for (const auto & [it, file_segment] : queue_entries)
+    for (auto & [it, file_segment] : queue_entries)
     {
-        /// Cell cache size changed and, for example, 1st file segment fits into cache
+        /// Cache size changed and, for example, 1st file segment fits into cache
         /// and 2nd file segment will fit only if first was evicted, then first will be removed and
-        /// cell is nullptr here.
+        /// file_segment_metadata is nullptr here.
         if (file_segment.expired())
             continue;
 
-        it->use(cache_lock);
+        LockedCachePriorityIterator(lock, it).use();
     }
-#ifndef NDEBUG
-    assertCacheCorrectness(cache_lock);
-#endif
 }
 
-void FileCache::reduceSizeToDownloaded(
-    const Key & key, size_t offset,
-    std::lock_guard<std::mutex> & cache_lock, std::unique_lock<std::mutex> & segment_lock)
+void FileCache::performDelayedRemovalOfDeletedKeysFromMetadata(const CacheMetadataGuard::Lock &)
 {
-    /**
-     * In case file was partially downloaded and it's download cannot be continued
-     * because of no space left in cache, we need to be able to cut cell's size to downloaded_size.
-     */
+    Key cleanup_key;
+    while (cleanup_keys_metadata_queue->tryPop(cleanup_key))
+    {
+        auto it = metadata.find(cleanup_key);
+        if (it == metadata.end())
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "No such key {} in metadata", cleanup_key.toString());
 
-    auto * cell = getCell(key, offset, cache_lock);
+        /// We must also lock the key.
+        auto guard = it->second->guard;
+        auto key_lock = guard->lock();
 
-    if (!cell)
-    {
-        throw Exception(
-            ErrorCodes::LOGICAL_ERROR,
-            "No cell found for key: {}, offset: {}",
-            key.toString(), offset);
+        /// Remove key from metadata.
+        metadata.erase(it);
+
+        try
+        {
+            /// Delete key directory if not yet deleted.
+            const fs::path path = fs::path(getPathInLocalCache(cleanup_key));
+            if (fs::exists(path))
+                fs::remove_all(path);
+
+            /// Delete three digit directory if empty.
+            const fs::path prefix_path = path.parent_path();
+            if (fs::exists(prefix_path) && fs::is_empty(prefix_path))
+                fs::remove_all(prefix_path);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+        }
     }
+}
 
-    const auto & file_segment = cell->file_segment;
+LockedKeyPtr FileCache::createLockedKey(const Key & key, KeyNotFoundPolicy key_not_found_policy)
+{
+    auto lock = metadata.lock();
+    performDelayedRemovalOfDeletedKeysFromMetadata(lock);
```

Comment:
> Which is only necessary here for calling 

no, this mutex is necessary not only for this reason, the main reason is actually because it guards a cache metadata (the std::unordered_map access to be more precise), which holds per key mutexes and allows to verify if the key is cached or not.

Reply:
Yes, got it wrong. But I'd say that the point still holds, since `performDelayedRemovalOfDeletedKeysFromMetadata()` does file's cleanup and this happened on hot path

Reply:
the reason why it is done on `createLockedKey` is explained here https://github.com/ClickHouse/ClickHouse/pull/44985/commits/2376f3a298e346dfae172bfe6302f4f101bac607#diff-8d9e69541877d0da116a7687e66fcc8b0dbd5454bd500e5978a2b5f6be8fdb53R950-R955.
Though it is possible to anyway avoid doing delayed removal here if we add some code for this exact case (checking for is_removed flag for only our key and update its state to non-removed). But on the other hand, this "delayed removal" is just removing some keys from std::unordered_map, which is lightweight, so may be we can neglect it?

Reply:
The file access on the hot path should be removed since it has non-predictable latency. We can do it in separate PR.

The idea is the following ... actually, the queue cleanup can be avoided here. We can safely create a new `KeyMetadata` for a key and place into `CacheMetadata`. If key metadata for the same key already exists and marked as removed, then just replace it. In this case, the cleanup queue still holds the pointer to old key metadata. 

When cleaning the queue in background, we need to check if `CacheMetadata` contains non-removed key metadata for the same key from the queue, - if so, then just drop the key metadata from the queue w/o removing the key folder. If it doesn't contain the key or contains the key marked as removed - it's safe to remove the key folder.